### PR TITLE
__delete shouldn't segfault for null

### PR DESCRIPTION
**Pull request by wilzbach**
See also: https://github.com/dlang/druntime/pull/2104#issuecomment-368101722

### CODE REVIEW

**Comment by schveiguy** [comment]
I think there is a different problem here!

This does not segfault:

```D
void main()
{
   int *x = null;
   GC.free(x);
}
```

File: src/core/memory.d [hunk 1]
```diff
@@ -1012,13 +1012,19 @@ void __delete(T)(ref T x) @system
...
+            x = null;
+        }
     }
     else static if (is(T == U2*, U2) || is(T : E2[], E2))
     {
-        GC.free(&x);
-        x = null;
+        if (x !is null)
+        {
+            GC.free(&x);
```
Comment (schveiguy) [inline src/core/memory.d]: This seems wrong, shouldn't it be `GC.free(x)` ?

Comment (wilzbach) [reply src/core/memory.d]: Yeah, looks like I was a bit too fast. Thanks!

**Comment by schveiguy** [comment]
OK, I see you fixed that problem. But now I'm slightly confused!

Because extracting out the problem:
```D
void main()
{
    int *ptr = null;
    GC.free(&ptr);
}
```

This segfaults, but this:

```D
void main()
{
   int *ptr = new int(1);
   GC.free(&ptr);
}
```

DOESN'T segfault. Neither is a valid heap pointer, so why does it segfault at all? The first thing `GC.free` should do is to find the pool the pointer is in, and it should fail in both cases.

In any case, a bug for another day. Your changes LGTM.

**Comment by schveiguy** [comment]
Hah, I think I know why it was segfaulting. Because of the new ProxyGC! Attempting to free a pointer before actually allocating any pointer will do this: https://github.com/dlang/druntime/blob/master/src/gc/impl/proto/gc.d#L140, which equates to a segfault when compiled in release mode.

**Comment by schveiguy** [comment]
This does fix a bug, though (`__delete` was not actually freeing the memory), so I'm pulling.

**Pull request merged**